fuzzers/libafl-fuzz/src/afl_stats.rs

```diff
@@ -0,0 +1,652 @@
+                "testcase must have scheduler metdata?",
+            ));
+        }
+        Ok(())
+    }
+
+    fn update_last_find(&mut self) {
+        self.last_find = current_time();
+    }
+
+    fn maybe_update_last_crash(
+        &mut self,
+        testcase: &Testcase<<<E as UsesState>::State as UsesInput>::Input>,
+        state: &E::State,
+    ) {
+        if testcase
+            .hit_objectives()
+            .contains(&Cow::Borrowed("CrashFeedback"))
+        {
+            self.last_crash = current_time();
+            self.execs_at_last_objective = *state.executions();
+        }
+    }
+
+    fn maybe_update_last_hang(
+        &mut self,
+        testcase: &Testcase<<<E as UsesState>::State as UsesInput>::Input>,
+        state: &E::State,
+    ) {
+        if testcase
+            .hit_objectives()
+            .contains(&Cow::Borrowed("TimeoutFeedback"))
+        {
+            self.last_hang = current_time();
+            self.execs_at_last_objective = *state.executions();
+        }
+    }
+
+    fn check_interval(&mut self) -> bool {
+        let cur = current_time();
+        if cur.checked_sub(self.last_report_time).unwrap_or_default() > self.stats_report_interval {
+            self.last_report_time = cur;
+            return true;
+        }
+        false
+    }
+    fn maybe_update_cycles(&mut self, queue_cycles: u64) {
+        if queue_cycles > self.cycles_done {
+            self.cycles_done += 1;
+        }
+    }
+
+    fn maybe_update_cycles_wo_finds(&mut self, queue_cycles: u64) {
+        if queue_cycles > self.cycles_done && self.last_find < current_time() {
+            self.cycles_wo_finds += 1;
+        }
+    }
+
+    #[allow(clippy::cast_precision_loss)]
+    #[allow(clippy::unused_self)]
+    fn calculate_stability(&self, unstable_entries: usize, filled_entries: u64) -> f64 {
+        ((filled_entries as f64 - unstable_entries as f64) / filled_entries as f64) * 100.0
+    }
+}
+
+impl Display for AFLPlotData<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(
+            f,
+            "{}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}",
+            self.relative_time,
+            self.cycles_done,
+            self.cur_item,
+            self.corpus_count,
+            self.pending_total,
+            self.pending_favs,
+            self.total_edges,
+            self.saved_crashes,
+            self.saved_hangs,
+            self.max_depth,
+            self.execs_per_sec,
+            self.execs_done,
+            self.edges_found
+        )
+    }
+}
+impl AFLPlotData<'_> {
+    fn get_header() -> String {
+        "# relative_time, cycles_done, cur_item, corpus_count, pending_total, pending_favs, total_edges, saved_crashes, saved_hangs, max_depth, execs_per_sec, execs_done, edges_found".to_string()
+    }
+}
+
+impl Display for AFLFuzzerStats<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```

Comment:
Can we somehow use serde for that? Feels wrong to do it by hand

Reply:
I checked and I don't think it's possible since we'd have to provide our own Serializer (not just implement Serialize) which defeats the purpose. I simplified the code though so should be more readable